<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18: http://docutils.sourceforge.net/" />
<meta property="og:title" content="Learning from unbalanced and small data" />
<meta property="og:type" content="website" />
<meta property="og:url" content="index/usage_principles/pu_learning.html" />
<meta property="og:site_name" content="AAanalysis" />
<meta property="og:description" content="In the life sciences, the prevalence of unbalanced and small datasets often leads to challenges in machine learning, particularly when dealing with a rare class of interest. Typically, binary class..." />
<meta property="og:image:width" content="1146" />
<meta property="og:image:height" content="600" />
<meta property="og:image" content="/_images/social_previews/summary_index_usage_principles_pu_learning_fc1a081b.png" />
<meta property="og:image:alt" content="In the life sciences, the prevalence of unbalanced and small datasets often leads to challenges in machine learning, particularly when dealing with a rare cl..." />
<meta name="description" content="In the life sciences, the prevalence of unbalanced and small datasets often leads to challenges in machine learning, particularly when dealing with a rare class of interest. Typically, binary class..." />
<meta name="twitter:card" content="summary_large_image" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Learning from unbalanced and small data &mdash; AAanalysis</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/plot_directive.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/design-style.1e8bd061cd6da7fc9cf755528e8ffc24.min.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/style.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/notebook.css" type="text/css" />
    <link rel="shortcut icon" href="../../_static/logo_small.png"/>
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/sphinx_highlight.js"></script>
        <script src="../../_static/clipboard.min.js"></script>
        <script src="../../_static/copybutton.js"></script>
        <script src="../../_static/design-tabs.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Explainable AI at Sequence Level" href="xai.html" />
    <link rel="prev" title="Identifying Physicochemical Signatures using CPP" href="feature_identification.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #343131" >

          
          
          <a href="../../index.html">
            
              <img src="../../_static/logo_big_trans.png" class="logo" alt="Logo"/>
          </a>
              <div class="version">
                latest
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">OVERVIEW</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Introduction</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../usage_principles.html">Usage Principles</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="data_flow_entry_points.html">Data Flow and Enry Points</a></li>
<li class="toctree-l2"><a class="reference internal" href="aaontology.html">AAontology: Classification of amino acid scales</a></li>
<li class="toctree-l2"><a class="reference internal" href="feature_identification.html">Identifying Physicochemical Signatures using CPP</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Learning from unbalanced and small data</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#what-is-pu-learning">What is PU learning?</a></li>
<li class="toctree-l3"><a class="reference internal" href="#challenges-and-potential-biases-in-pu-learning">Challenges and Potential Biases in PU Learning</a></li>
<li class="toctree-l3"><a class="reference internal" href="#evaluation-strategies-for-pu-learning">Evaluation Strategies for PU Learning</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="xai.html">Explainable AI at Sequence Level</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../CONTRIBUTING_COPY.html">Contributing</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">EXAMPLES</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../tutorials.html">Tutorials</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">REFERENCES</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../api.html">API</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../tables.html">Tables</a></li>
<li class="toctree-l1"><a class="reference internal" href="../references.html">References</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #343131" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">AAanalysis</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../usage_principles.html">Usage Principles</a></li>
      <li class="breadcrumb-item active">Learning from unbalanced and small data</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/breimanntools/aaanalysis/blob/master/index/usage_principles/pu_learning.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="learning-from-unbalanced-and-small-data">
<h1>Learning from unbalanced and small data<a class="headerlink" href="#learning-from-unbalanced-and-small-data" title="Permalink to this heading"></a></h1>
<p>In the life sciences, the prevalence of unbalanced and small datasets often leads to challenges in machine learning,
particularly when dealing with a rare class of interest. Typically, binary classification involves labeled positive
and negative data. However, in scenarios where negative samples are scarce, conventional data augmentation methods
like SMOTE may fall short due to generating artificial samples. This issue is especially pertinent in fields like
protein sequence prediction, where even minor alterations can lead to significant biological implications.
A practical solution is to harness the abundance of unlabeled data to identify negative samples, a strategy
that becomes essential when obtaining a balanced dataset is challenging.</p>
<section id="what-is-pu-learning">
<h2>What is PU learning?<a class="headerlink" href="#what-is-pu-learning" title="Permalink to this heading"></a></h2>
<p>Positive-Unlabeled (PU) learning, a subfield of machine learning, is tailored for situations with only positive and
unlabeled data. It is gaining relevance in bioinformatics and similar fields, where negative data often remain
unlabeled and undiscovered. PU learning algorithms aim to identify negative data from unlabeled data based on statistical
comparison with the positive data, coupled with iterative learning strategies. This approach adeptly handles the
inherent data asymmetry in applications where labeled negatives are unattainable.</p>
</section>
<section id="challenges-and-potential-biases-in-pu-learning">
<h2>Challenges and Potential Biases in PU Learning<a class="headerlink" href="#challenges-and-potential-biases-in-pu-learning" title="Permalink to this heading"></a></h2>
<p>PU learning faces specific challenges, that can be summarized as follows:</p>
<ul class="simple">
<li><p><strong>Homogeneity in Identified Negatives</strong>: Similarity among identified negatives can indicate algorithm bias or
training data diversity issues, potentially leading to overfitting.</p></li>
<li><p><strong>Biased Negative Selection</strong>: Selection methods for negatives can significantly influence model performance.</p></li>
<li><p><strong>Overfitting to Positive Data</strong>: Training predominantly on positive data risks failing to accurately identify true negatives.</p></li>
<li><p><strong>Assumptions in Class Distribution</strong>: Incorrect assumptions about class distribution can result in misclassification.</p></li>
<li><p><strong>Data Quality</strong>: The quality and representativeness of data are critical for model accuracy.</p></li>
</ul>
<p>To enhance model robustness and accuracy, it is crucial to ensure a high diversity among identified negatives. This
diversity promotes better generalization, as models learn to recognize a broader spectrum of negative features.
Additionally, negatives near the decision boundary are particularly valuable for the model’s ability to distinguish
between positives and negatives. Addressing these challenges requires careful algorithm selection, data preprocessing,
and diverse evaluation strategies.</p>
</section>
<section id="evaluation-strategies-for-pu-learning">
<h2>Evaluation Strategies for PU Learning<a class="headerlink" href="#evaluation-strategies-for-pu-learning" title="Permalink to this heading"></a></h2>
<p>Evaluating PU learning models in the absence of ground truth can be performed in two steps. First, the population
of identified negatives can be assessed regarding their statistical properties and compared against the other data
classes. Second, the prediction performance of the machine learning models trained on the obtained dataset can be assessed.
The availability of a minority ground-truth negative dataset can significantly aid this process.</p>
<p><strong>1. Statistical Evaluation of Identified Negatives</strong>
This step involves analyzing the characteristics of the negatives identified by the PU learning model using
distinct statistical measures:</p>
<ul class="simple">
<li><p><strong>Homogeneity in Identified Negatives</strong>: Examine the extent to which identified negatives are similar
to each other, which can indicate potential biases or lack of diversity in the training data. This can be achieved
by using:</p>
<ul>
<li><p><em>Coefficient of Variation (CV)</em> [0, ∞): Evaluates the variability within the identified negatives,
with high CV (&gt; 1) indicating more diversity and low CV (&lt; 1) indicating homogeneity</p></li>
<li><p><em>Entropy</em> [0, 1]: Measures the diversity within the identified negatives, where higher entropy signifies greater
variability and lower entropy indicates homogeneity</p></li>
</ul>
</li>
<li><p><strong>Distribution Alignment</strong>: Determine the similarity between the identified negatives and the other data classes,
by assessing the alignment of their distributions. This is especially useful if a minority ground-truth dataset
of negatives exists. To this end, various statistical measures can be used:</p>
<ul>
<li><p><em>Area Under the Curve (AUC)</em>: Assess the difference between two datasets, with a higher AUC indicating
a higher discrimination. We use an adjusted AUC [-0.5, 0.5] such that 0 indicates no differences and the
sign indicates which dataset comprises higher values.</p></li>
<li><p><em>Kullback-Leibler Divergence</em> [0, ∞): Measures the divergence between the distributions of identified negatives
and unlabeled data, with lower values (near 0) indicating better alignment.</p></li>
<li><p><em>Mann-Whitney U Test</em>: Non-parametric test to compare distributions, useful for assessing if the identified negatives
differ significantly from the positive or unlabeled data.</p></li>
</ul>
</li>
<li><p><strong>Assessing Reproducibility</strong>: To evaluate the reproducibility of a PU learning algorithm, perform the method multiple
times and compare the overlap in the resulting sets of identified negatives.</p></li>
</ul>
<p><strong>2. Machine Learning Model Evaluation Strategies</strong>
Once the negatives are identified, the following strategies can be employed to evaluate the trained models using
standard evaluation measures such as accuracy or balanced accuracy (preferred if dataset is still unbalanced):</p>
<ul class="simple">
<li><p><strong>Proxy Ground Truth</strong>: Utilize a small, accurately labeled dataset as a proxy for ground truth to provide
a limited yet insightful evaluation. If a minority ground truth negative dataset is available, you can train only
using the known positive and identified negative samples, and evaluate on the ground truth negative data.</p></li>
<li><p><strong>Consistency and Stability</strong>: Assess the model’s consistency across multiple runs or data subsets,
indicating reliability.</p></li>
<li><p><strong>Confidence Analysis</strong>: Evaluate the model’s prediction confidence scores. High confidence may indicate effectiveness,
but beware of overfitting or bias.</p></li>
<li><p><strong>External Validation</strong>: Use external knowledge or domain expertise to validate model predictions,
ensuring they align with real-world scenarios.</p></li>
<li><p><strong>Ablation and Sensitivity Analysis</strong>: Conduct ablation studies to understand model dependencies, and test the
model’s robustness to data perturbations. You can assess the impact of removing or altering features or parameters,
indicating model robustness.</p></li>
</ul>
<p>Combining these strategies offers a comprehensive view of a PU learning model’s performance, especially in scenarios
lacking complete ground truth data.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="feature_identification.html" class="btn btn-neutral float-left" title="Identifying Physicochemical Signatures using CPP" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="xai.html" class="btn btn-neutral float-right" title="Explainable AI at Sequence Level" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Stephan Breimann.</p>
  </div>

   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>