<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18: http://docutils.sourceforge.net/" />
<meta property="og:title" content="Quick Start with AAanalysis" />
<meta property="og:type" content="website" />
<meta property="og:url" content="generated/tutorial1_quick_start.html" />
<meta property="og:site_name" content="AAanalysis" />
<meta property="og:description" content="Dive into the powerful capabilities of AAanalysis —a Python framework dedicated to sequence-based, alignment-free protein prediction. In this tutorial, using gamma-secretase substrates and non-subs..." />
<meta property="og:image:width" content="1146" />
<meta property="og:image:height" content="600" />
<meta property="og:image" content="/_images/social_previews/summary_generated_tutorial1_quick_start_8b3b7a01.png" />
<meta property="og:image:alt" content="Dive into the powerful capabilities of AAanalysis —a Python framework dedicated to sequence-based, alignment-free protein prediction. In this tutorial, using..." />
<meta name="description" content="Dive into the powerful capabilities of AAanalysis —a Python framework dedicated to sequence-based, alignment-free protein prediction. In this tutorial, using gamma-secretase substrates and non-subs..." />
<meta name="twitter:card" content="summary_large_image" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Quick Start with AAanalysis &mdash; AAanalysis</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="../_static/plot_directive.css" type="text/css" />
      <link rel="stylesheet" href="../_static/design-style.1e8bd061cd6da7fc9cf755528e8ffc24.min.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/style.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/notebook.css" type="text/css" />
    <link rel="shortcut icon" href="../_static/logo_small.png"/>
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
        <script src="../_static/clipboard.min.js"></script>
        <script src="../_static/copybutton.js"></script>
        <script src="../_static/design-tabs.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Plotting prelude" href="plotting_prelude.html" />
    <link rel="prev" title="Tutorials" href="../tutorials.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #343131" >

          
          
          <a href="../index.html">
            
              <img src="../_static/logo_big_trans.png" class="logo" alt="Logo"/>
          </a>
              <div class="version">
                latest
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">OVERVIEW</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../index/introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../index/usage_principles.html">Usage Principles</a></li>
<li class="toctree-l1"><a class="reference internal" href="../index/CONTRIBUTING_COPY.html">Contributing</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">EXAMPLES</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../tutorials.html">Tutorials</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../tutorials.html#quick-start">Quick start</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">Quick Start with AAanalysis</a></li>
<li class="toctree-l3"><a class="reference internal" href="plotting_prelude.html">Plotting prelude</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials.html#further-tutorials">Further Tutorials</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">REFERENCES</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../api.html">API</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../index/tables.html">Tables</a></li>
<li class="toctree-l1"><a class="reference internal" href="../index/references.html">References</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #343131" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">AAanalysis</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../tutorials.html">Tutorials</a></li>
      <li class="breadcrumb-item active">Quick Start with AAanalysis</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/breimanntools/aaanalysis/blob/master/generated/tutorial1_quick_start.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="quick-start-with-aaanalysis">
<h1>Quick Start with AAanalysis<a class="headerlink" href="#quick-start-with-aaanalysis" title="Permalink to this heading"></a></h1>
<p>Dive into the powerful capabilities of <code class="docutils literal notranslate"><span class="pre">AAanalysis</span></code>—a Python framework
dedicated to sequence-based, alignment-free protein prediction. In this
tutorial, using gamma-secretase substrates and non-substrates as an
example, we’ll focus on extracting interpretable features from protein
sequences using the <code class="docutils literal notranslate"><span class="pre">AAclust</span></code> and <code class="docutils literal notranslate"><span class="pre">CPP</span></code> models and how they can be
harnessed for binary classification tasks.</p>
<section id="what-you-will-learn">
<h2>What You Will Learn:<a class="headerlink" href="#what-you-will-learn" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">Loading</span> <span class="pre">Sequences</span> <span class="pre">and</span> <span class="pre">Scales</span></code>: How to easily load protein
sequences and their amino acid scales.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Feature</span> <span class="pre">Engineering</span></code>: Extract essential features using the
<code class="docutils literal notranslate"><span class="pre">AAclust</span></code> and <code class="docutils literal notranslate"><span class="pre">CPP</span></code> models.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Protein</span> <span class="pre">Prediction</span></code>: Make predictions using the RandomForest
model.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Explainable</span> <span class="pre">AI</span></code>: Interpret predictions at the group and individual
levels by combining <code class="docutils literal notranslate"><span class="pre">CPP</span></code> with <code class="docutils literal notranslate"><span class="pre">SHAP</span></code>.</p></li>
</ul>
</section>
<section id="loading-sequences-and-scales">
<h2>1. Loading Sequences and Scales<a class="headerlink" href="#loading-sequences-and-scales" title="Permalink to this heading"></a></h2>
<p>With AAanalysis, you have access to numerous benchmark datasets for
protein sequence analysis. Using our γ-secretase substrates and
non-substrates dataset as a hands-on example, you can effortlessly
retrieve these datasets using the <code class="docutils literal notranslate"><span class="pre">aa.load_dataset()</span></code> function.
Furthermore, amino acid scales, predominantly from AAindex, along with
their hierarchical classification (known as <code class="docutils literal notranslate"><span class="pre">AAontology</span></code>), are
available at your fingertips with the <code class="docutils literal notranslate"><span class="pre">aa.load_scales()</span></code> function.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">aaanalysis</span> <span class="k">as</span> <span class="nn">aa</span>

<span class="n">df_scales</span> <span class="o">=</span> <span class="n">aa</span><span class="o">.</span><span class="n">load_scales</span><span class="p">()</span>
<span class="n">df_seq</span> <span class="o">=</span> <span class="n">aa</span><span class="o">.</span><span class="n">load_dataset</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;DOM_GSEC&quot;</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>
<span class="n">df_seq</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>entry</th>
      <th>sequence</th>
      <th>label</th>
      <th>tmd_start</th>
      <th>tmd_stop</th>
      <th>jmd_n</th>
      <th>tmd</th>
      <th>jmd_c</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Q14802</td>
      <td>MQKVTLGLLVFLAGFPVLDANDLEDKNSPFYYDWHSLQVGGLICAG...</td>
      <td>0</td>
      <td>37</td>
      <td>59</td>
      <td>NSPFYYDWHS</td>
      <td>LQVGGLICAGVLCAMGIIIVMSA</td>
      <td>KCKCKFGQKS</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Q86UE4</td>
      <td>MAARSWQDELAQQAEEGSARLREMLSVGLGFLRTELGLDLGLEPKR...</td>
      <td>0</td>
      <td>50</td>
      <td>72</td>
      <td>LGLEPKRYPG</td>
      <td>WVILVGTGALGLLLLFLLGYGWA</td>
      <td>AACAGARKKR</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Q969W9</td>
      <td>MHRLMGVNSTAAAAAGQPNVSCTCNCKRSLFQSMEITELEFVQIII...</td>
      <td>0</td>
      <td>41</td>
      <td>63</td>
      <td>FQSMEITELE</td>
      <td>FVQIIIIVVVMMVMVVVITCLLS</td>
      <td>HYKLSARSFI</td>
    </tr>
    <tr>
      <th>3</th>
      <td>P53801</td>
      <td>MAPGVARGPTPYWRLRLGGAALLLLLIPVAAAQEPPGAACSQNTNK...</td>
      <td>0</td>
      <td>97</td>
      <td>119</td>
      <td>RWGVCWVNFE</td>
      <td>ALIITMSVVGGTLLLGIAICCCC</td>
      <td>CCRRKRSRKP</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Q8IUW5</td>
      <td>MAPRALPGSAVLAAAVFVGGAVSSPLVAPDNGSSRTLHSRTETTPS...</td>
      <td>0</td>
      <td>59</td>
      <td>81</td>
      <td>NDTGNGHPEY</td>
      <td>IAYALVPVFFIMGLFGVLICHLL</td>
      <td>KKKGYRCTTE</td>
    </tr>
  </tbody>
</table>
</div></section>
<section id="feature-engineering">
<h2>2. Feature Engineering<a class="headerlink" href="#feature-engineering" title="Permalink to this heading"></a></h2>
<p>The centerpiece of AAanalysis is the Comparative Physicochemical
Profiling (<code class="docutils literal notranslate"><span class="pre">CPP</span></code>) model, which is supported by <code class="docutils literal notranslate"><span class="pre">AAclust</span></code> for the
pre-selection of amino acid scales.</p>
<section id="aaclust">
<h3>AAclust<a class="headerlink" href="#aaclust" title="Permalink to this heading"></a></h3>
<p>Since redundancy is an essential problem for machine learning tasks, the
<code class="docutils literal notranslate"><span class="pre">AAclust</span></code> object provides a lightweight wrapper for sklearn clustering
algorithms such as Agglomerative clustering. AAclust clusters a set of
scales and selects for each cluster the most representative scale (i.e.,
the scale closes to the cluster center). We will use AAclust to obtain a
set of 100 scales, as defined by the <code class="docutils literal notranslate"><span class="pre">n_clusters</span></code> parameters:</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">sklearn.cluster</span> <span class="kn">import</span> <span class="n">AgglomerativeClustering</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="n">aac</span> <span class="o">=</span> <span class="n">aa</span><span class="o">.</span><span class="n">AAclust</span><span class="p">(</span><span class="n">model</span><span class="o">=</span><span class="n">AgglomerativeClustering</span><span class="p">)</span>
<span class="n">X</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">df_scales</span><span class="p">)</span>
<span class="n">scales</span> <span class="o">=</span> <span class="n">aac</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">df_scales</span><span class="p">),</span> <span class="n">n_clusters</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">df_scales</span> <span class="o">=</span> <span class="n">df_scales</span><span class="p">[</span><span class="n">scales</span><span class="p">]</span>
<span class="n">df_scales</span><span class="p">[</span><span class="n">scales</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">4</span><span class="p">]]</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ANDN920101</th>
      <th>SIMZ760101</th>
      <th>NAKH900106</th>
      <th>AURR980112</th>
    </tr>
    <tr>
      <th>AA</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>A</th>
      <td>0.494</td>
      <td>0.268</td>
      <td>0.237</td>
      <td>0.787</td>
    </tr>
    <tr>
      <th>C</th>
      <td>0.864</td>
      <td>0.258</td>
      <td>0.303</td>
      <td>0.104</td>
    </tr>
    <tr>
      <th>D</th>
      <td>1.000</td>
      <td>0.206</td>
      <td>0.000</td>
      <td>0.451</td>
    </tr>
    <tr>
      <th>E</th>
      <td>0.420</td>
      <td>0.210</td>
      <td>0.090</td>
      <td>0.823</td>
    </tr>
    <tr>
      <th>F</th>
      <td>0.877</td>
      <td>0.887</td>
      <td>0.724</td>
      <td>0.402</td>
    </tr>
  </tbody>
</table>
</div></section>
<section id="comparative-physicochemical-profiling-cpp">
<h3>Comparative Physicochemical Profiling (CPP)<a class="headerlink" href="#comparative-physicochemical-profiling-cpp" title="Permalink to this heading"></a></h3>
<p>CPP is a sequence-based feature engineering algorithm. It aims at
identifying a set of features most discriminant between two sets of
sequences: the test set and the reference set. Supported by the
<code class="docutils literal notranslate"><span class="pre">SequenceFeature</span></code> object (<code class="docutils literal notranslate"><span class="pre">sf</span></code>), A CPP feature integrates: -
<code class="docutils literal notranslate"><span class="pre">Parts</span></code>: Are combination of a target middle domain (TMD) and N- and
C-terminal adjacent regions (JMD-N and JMD-C, respectively), obtained
<code class="docutils literal notranslate"><span class="pre">sf.get_df_parts</span></code>. - <code class="docutils literal notranslate"><span class="pre">Splits</span></code>: These <code class="docutils literal notranslate"><span class="pre">Parts</span></code> can be split into
various continuous segments or discontinuous patterns, specified
<code class="docutils literal notranslate"><span class="pre">sf.get_split_kws()</span></code>. - <code class="docutils literal notranslate"><span class="pre">Scales</span></code>: Sets of amino acid scales.</p>
<p>We use SequenceFeature to obtain Parts and Splits:</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">y</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">df_seq</span><span class="p">[</span><span class="s2">&quot;label&quot;</span><span class="p">])</span>
<span class="n">sf</span> <span class="o">=</span> <span class="n">aa</span><span class="o">.</span><span class="n">SequenceFeature</span><span class="p">()</span>
<span class="n">df_parts</span> <span class="o">=</span> <span class="n">sf</span><span class="o">.</span><span class="n">get_df_parts</span><span class="p">(</span><span class="n">df_seq</span><span class="o">=</span><span class="n">df_seq</span><span class="p">,</span> <span class="n">list_parts</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;tmd_jmd&quot;</span><span class="p">])</span>
<span class="n">split_kws</span> <span class="o">=</span> <span class="n">sf</span><span class="o">.</span><span class="n">get_split_kws</span><span class="p">(</span><span class="n">n_split_max</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">split_types</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Segment&quot;</span><span class="p">])</span>
<span class="n">df_parts</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>tmd_jmd</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>D3ZZK3</th>
      <td>RIIGDGANSTVLLVSVSGSVVLVVILIAAFVISRRRSKYSQAK</td>
    </tr>
    <tr>
      <th>O14786</th>
      <td>PGNVLKTLDPILITIIAMSALGVLLGAVCGVVLYCACWHNGMS</td>
    </tr>
    <tr>
      <th>O35516</th>
      <td>SELESPRNAQLLYLLAVAVVIILFFILLGVIMAKRKRKHGFLW</td>
    </tr>
    <tr>
      <th>O43914</th>
      <td>DCSCSTVSPGVLAGIVMGDLVLTVLIALAVYFLGRLVPRGRGA</td>
    </tr>
    <tr>
      <th>O75581</th>
      <td>YPTEEPAPQATNTVGSVIGVIVTIFVSGTVYFICQRMLCPRMK</td>
    </tr>
  </tbody>
</table>
</div><p>Running the CPP algorithm creates all <code class="docutils literal notranslate"><span class="pre">Part</span></code>, <code class="docutils literal notranslate"><span class="pre">Split</span></code>, <code class="docutils literal notranslate"><span class="pre">Split</span></code>
combinations and filters a selected maximum of non-redundant features.
As a baseline approach, we use CPP to compute the average values for the
100 selected scales over the entire TMD-JMD sequences:</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Small set of CPP features (100 features are created)</span>
<span class="n">cpp</span> <span class="o">=</span> <span class="n">aa</span><span class="o">.</span><span class="n">CPP</span><span class="p">(</span><span class="n">df_scales</span><span class="o">=</span><span class="n">df_scales</span><span class="p">,</span> <span class="n">df_parts</span><span class="o">=</span><span class="n">df_parts</span><span class="p">,</span> <span class="n">split_kws</span><span class="o">=</span><span class="n">split_kws</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">df_feat</span> <span class="o">=</span> <span class="n">cpp</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">labels</span><span class="o">=</span><span class="n">y</span><span class="p">)</span>
<span class="n">df_feat</span>
</pre></div>
</div>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>feature</th>
      <th>category</th>
      <th>subcategory</th>
      <th>scale_name</th>
      <th>scale_description</th>
      <th>abs_auc</th>
      <th>abs_mean_dif</th>
      <th>mean_dif</th>
      <th>std_test</th>
      <th>std_ref</th>
      <th>p_val_mann_whitney</th>
      <th>p_val_fdr_bh</th>
      <th>positions</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>TMD_JMD-Segment(1,1)-ANDN920101</td>
      <td>Structure-Activity</td>
      <td>Backbone-dynamics (-CH)</td>
      <td>α-CH chemical shifts (backbone-dynamics)</td>
      <td>alpha-CH chemical shifts (Andersen et al., 1992)</td>
      <td>0.130</td>
      <td>0.022966</td>
      <td>0.022966</td>
      <td>0.054433</td>
      <td>0.053266</td>
      <td>0.025737</td>
      <td>0.099022</td>
      <td>1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...</td>
    </tr>
    <tr>
      <th>1</th>
      <td>TMD_JMD-Segment(1,1)-VASM830101</td>
      <td>Conformation</td>
      <td>Unclassified (Conformation)</td>
      <td>α-helix</td>
      <td>Relative population of conformational state A ...</td>
      <td>0.120</td>
      <td>0.019298</td>
      <td>-0.019298</td>
      <td>0.046755</td>
      <td>0.049127</td>
      <td>0.039609</td>
      <td>0.099022</td>
      <td>1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...</td>
    </tr>
    <tr>
      <th>2</th>
      <td>TMD_JMD-Segment(1,1)-ROBB760113</td>
      <td>Conformation</td>
      <td>β-turn</td>
      <td>β-turn</td>
      <td>Information measure for loop (Robson-Suzuki, 1...</td>
      <td>0.108</td>
      <td>0.021958</td>
      <td>0.021958</td>
      <td>0.060658</td>
      <td>0.053190</td>
      <td>0.062212</td>
      <td>0.100670</td>
      <td>1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...</td>
    </tr>
    <tr>
      <th>3</th>
      <td>TMD_JMD-Segment(1,1)-RACS820103</td>
      <td>Conformation</td>
      <td>Unclassified (Conformation)</td>
      <td>α-helix (left-handed)</td>
      <td>Average relative fractional occurrence in AL(i...</td>
      <td>0.080</td>
      <td>0.019579</td>
      <td>-0.019579</td>
      <td>0.072260</td>
      <td>0.047452</td>
      <td>0.166907</td>
      <td>0.166907</td>
      <td>1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...</td>
    </tr>
  </tbody>
</table>
</div></section>
</section>
<section id="protein-prediction">
<h2>3. Protein Prediction<a class="headerlink" href="#protein-prediction" title="Permalink to this heading"></a></h2>
<p>A feature matrix from a given set of CPP features can be created using
<code class="docutils literal notranslate"><span class="pre">sf.feat_matrix</span></code> and used for machine learning:</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">sklearn.ensemble</span> <span class="kn">import</span> <span class="n">RandomForestClassifier</span>
<span class="kn">from</span> <span class="nn">sklearn.model_selection</span> <span class="kn">import</span> <span class="n">cross_val_score</span>

<span class="n">X</span> <span class="o">=</span> <span class="n">sf</span><span class="o">.</span><span class="n">feat_matrix</span><span class="p">(</span><span class="n">df_parts</span><span class="o">=</span><span class="n">df_parts</span><span class="p">,</span> <span class="n">features</span><span class="o">=</span><span class="n">df_feat</span><span class="p">[</span><span class="s2">&quot;feature&quot;</span><span class="p">])</span>
<span class="n">rf</span> <span class="o">=</span> <span class="n">RandomForestClassifier</span><span class="p">()</span>
<span class="n">cv_base</span> <span class="o">=</span> <span class="n">cross_val_score</span><span class="p">(</span><span class="n">rf</span><span class="p">,</span> <span class="n">X</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">scoring</span><span class="o">=</span><span class="s2">&quot;accuracy&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Mean accuracy of </span><span class="si">{</span><span class="nb">round</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">cv_base</span><span class="p">),</span><span class="w"> </span><span class="mi">2</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Mean</span> <span class="n">accuracy</span> <span class="n">of</span> <span class="mf">0.58</span>
</pre></div>
</div>
<p>Creating more features with CPP will take some more time. but improve
prediction performance:</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># CPP features with default splits (around 100.000 features)</span>
<span class="n">df_parts</span> <span class="o">=</span> <span class="n">sf</span><span class="o">.</span><span class="n">get_df_parts</span><span class="p">(</span><span class="n">df_seq</span><span class="o">=</span><span class="n">df_seq</span><span class="p">)</span>
<span class="n">cpp</span> <span class="o">=</span> <span class="n">aa</span><span class="o">.</span><span class="n">CPP</span><span class="p">(</span><span class="n">df_scales</span><span class="o">=</span><span class="n">df_scales</span><span class="p">,</span> <span class="n">df_parts</span><span class="o">=</span><span class="n">df_parts</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">df_feat</span> <span class="o">=</span> <span class="n">cpp</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">labels</span><span class="o">=</span><span class="n">y</span><span class="p">)</span>
<span class="n">df_feat</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>feature</th>
      <th>category</th>
      <th>subcategory</th>
      <th>scale_name</th>
      <th>scale_description</th>
      <th>abs_auc</th>
      <th>abs_mean_dif</th>
      <th>mean_dif</th>
      <th>std_test</th>
      <th>std_ref</th>
      <th>p_val_mann_whitney</th>
      <th>p_val_fdr_bh</th>
      <th>positions</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>TMD_C_JMD_C-Segment(2,3)-QIAN880106</td>
      <td>Conformation</td>
      <td>α-helix</td>
      <td>α-helix (middle)</td>
      <td>Weights for alpha-helix at the window position...</td>
      <td>0.387</td>
      <td>0.121446</td>
      <td>0.121446</td>
      <td>0.069196</td>
      <td>0.085013</td>
      <td>0.000000e+00</td>
      <td>0.000000e+00</td>
      <td>27,28,29,30,31,32,33</td>
    </tr>
    <tr>
      <th>1</th>
      <td>TMD_C_JMD_C-Segment(4,5)-ZIMJ680104</td>
      <td>Energy</td>
      <td>Isoelectric point</td>
      <td>Isoelectric point</td>
      <td>Isoelectric point (Zimmerman et al., 1968)</td>
      <td>0.373</td>
      <td>0.220000</td>
      <td>0.220000</td>
      <td>0.123716</td>
      <td>0.137350</td>
      <td>1.000000e-10</td>
      <td>2.475000e-07</td>
      <td>33,34,35,36</td>
    </tr>
    <tr>
      <th>2</th>
      <td>TMD_C_JMD_C-Pattern(N,5,8,12,15)-QIAN880106</td>
      <td>Conformation</td>
      <td>α-helix</td>
      <td>α-helix (middle)</td>
      <td>Weights for alpha-helix at the window position...</td>
      <td>0.358</td>
      <td>0.144860</td>
      <td>0.144860</td>
      <td>0.079321</td>
      <td>0.117515</td>
      <td>7.000000e-10</td>
      <td>7.150000e-07</td>
      <td>25,28,32,35</td>
    </tr>
    <tr>
      <th>3</th>
      <td>TMD_C_JMD_C-Segment(5,7)-LINS030101</td>
      <td>ASA/Volume</td>
      <td>Volume</td>
      <td>Accessible surface area (ASA)</td>
      <td>Total accessible surfaces of whole residues (b...</td>
      <td>0.354</td>
      <td>0.237161</td>
      <td>0.237161</td>
      <td>0.145884</td>
      <td>0.164285</td>
      <td>1.100000e-09</td>
      <td>7.150000e-07</td>
      <td>32,33,34</td>
    </tr>
    <tr>
      <th>4</th>
      <td>TMD_C_JMD_C-Segment(6,9)-ZIMJ680104</td>
      <td>Energy</td>
      <td>Isoelectric point</td>
      <td>Isoelectric point</td>
      <td>Isoelectric point (Zimmerman et al., 1968)</td>
      <td>0.341</td>
      <td>0.263651</td>
      <td>0.263651</td>
      <td>0.187136</td>
      <td>0.171995</td>
      <td>4.000000e-09</td>
      <td>1.185395e-06</td>
      <td>32,33</td>
    </tr>
    <tr>
      <th>5</th>
      <td>TMD_C_JMD_C-Segment(4,9)-ROBB760113</td>
      <td>Conformation</td>
      <td>β-turn</td>
      <td>β-turn</td>
      <td>Information measure for loop (Robson-Suzuki, 1...</td>
      <td>0.337</td>
      <td>0.319440</td>
      <td>-0.319440</td>
      <td>0.175203</td>
      <td>0.255754</td>
      <td>6.100000e-09</td>
      <td>1.185395e-06</td>
      <td>27,28</td>
    </tr>
    <tr>
      <th>6</th>
      <td>TMD_C_JMD_C-Segment(2,2)-EISD860102</td>
      <td>Energy</td>
      <td>Isoelectric point</td>
      <td>Atom-based hydrophobic moment</td>
      <td>Atom-based hydrophobic moment (Eisenberg-McLac...</td>
      <td>0.337</td>
      <td>0.139567</td>
      <td>0.139567</td>
      <td>0.098917</td>
      <td>0.101842</td>
      <td>6.300000e-09</td>
      <td>1.185395e-06</td>
      <td>31,32,33,34,35,36,37,38,39,40</td>
    </tr>
    <tr>
      <th>7</th>
      <td>TMD_C_JMD_C-Segment(4,5)-RICJ880113</td>
      <td>Conformation</td>
      <td>α-helix (C-cap)</td>
      <td>α-helix (C-terminal, inside)</td>
      <td>Relative preference value at C2 (Richardson-Ri...</td>
      <td>0.336</td>
      <td>0.223765</td>
      <td>0.223765</td>
      <td>0.133513</td>
      <td>0.178217</td>
      <td>7.100000e-09</td>
      <td>1.185395e-06</td>
      <td>33,34,35,36</td>
    </tr>
    <tr>
      <th>8</th>
      <td>TMD_C_JMD_C-Segment(5,7)-KARS160107</td>
      <td>Shape</td>
      <td>Side chain length</td>
      <td>Eccentricity (maximum)</td>
      <td>Diameter (maximum eccentricity) (Karkbara-Knis...</td>
      <td>0.331</td>
      <td>0.217594</td>
      <td>0.217594</td>
      <td>0.136011</td>
      <td>0.172395</td>
      <td>1.130000e-08</td>
      <td>1.331786e-06</td>
      <td>32,33,34</td>
    </tr>
    <tr>
      <th>9</th>
      <td>TMD_C_JMD_C-Pattern(C,4,8)-JURD980101</td>
      <td>Polarity</td>
      <td>Hydrophobicity</td>
      <td>Hydrophobicity</td>
      <td>Modified Kyte-Doolittle hydrophobicity scale (...</td>
      <td>0.329</td>
      <td>0.264720</td>
      <td>-0.264720</td>
      <td>0.141666</td>
      <td>0.233134</td>
      <td>1.480000e-08</td>
      <td>1.425259e-06</td>
      <td>33,37</td>
    </tr>
  </tbody>
</table>
</div><p>Which can be again used for machine learning:</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">simplefilter</span><span class="p">(</span><span class="n">action</span><span class="o">=</span><span class="s1">&#39;ignore&#39;</span><span class="p">,</span> <span class="n">category</span><span class="o">=</span><span class="ne">FutureWarning</span><span class="p">)</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">X</span> <span class="o">=</span> <span class="n">sf</span><span class="o">.</span><span class="n">feat_matrix</span><span class="p">(</span><span class="n">df_parts</span><span class="o">=</span><span class="n">df_parts</span><span class="p">,</span> <span class="n">features</span><span class="o">=</span><span class="n">df_feat</span><span class="p">[</span><span class="s2">&quot;feature&quot;</span><span class="p">])</span>
<span class="n">rf</span> <span class="o">=</span> <span class="n">RandomForestClassifier</span><span class="p">()</span>
<span class="n">cv</span> <span class="o">=</span> <span class="n">cross_val_score</span><span class="p">(</span><span class="n">rf</span><span class="p">,</span> <span class="n">X</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">scoring</span><span class="o">=</span><span class="s2">&quot;accuracy&quot;</span><span class="p">,</span> <span class="n">cv</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Mean accuracy of </span><span class="si">{</span><span class="nb">round</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">cv</span><span class="p">),</span><span class="w"> </span><span class="mi">2</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="n">aa</span><span class="o">.</span><span class="n">plot_settings</span><span class="p">(</span><span class="n">font_scale</span><span class="o">=</span><span class="mf">1.1</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;Baseline&quot;</span><span class="p">:</span> <span class="n">cv_base</span><span class="p">,</span> <span class="s2">&quot;CPP&quot;</span><span class="p">:</span> <span class="n">cv</span><span class="p">}),</span> <span class="n">palette</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;tab:blue&quot;</span><span class="p">,</span> <span class="s2">&quot;tab:red&quot;</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Mean accuracy&quot;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">aa</span><span class="o">.</span><span class="n">plot_gcfs</span><span class="p">()</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">despine</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Mean</span> <span class="n">accuracy</span> <span class="n">of</span> <span class="mf">0.9</span>
</pre></div>
</div>
<img alt="../_images/output_13_1.png" src="../_images/output_13_1.png" />
</section>
<section id="explainable-ai">
<h2>4. Explainable AI<a class="headerlink" href="#explainable-ai" title="Permalink to this heading"></a></h2>
<section id="explainable-ai-on-group-level">
<h3>Explainable AI on group level<a class="headerlink" href="#explainable-ai-on-group-level" title="Permalink to this heading"></a></h3>
</section>
<section id="explainable-ai-on-individual-level">
<h3>Explainable AI on individual level<a class="headerlink" href="#explainable-ai-on-individual-level" title="Permalink to this heading"></a></h3>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../tutorials.html" class="btn btn-neutral float-left" title="Tutorials" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="plotting_prelude.html" class="btn btn-neutral float-right" title="Plotting prelude" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Stephan Breimann.</p>
  </div>

   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>