<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18: http://docutils.sourceforge.net/" />
<meta property="og:title" content="Slow Start with AAanalysis" />
<meta property="og:type" content="website" />
<meta property="og:url" content="generated/tutorial1_slow_start.html" />
<meta property="og:site_name" content="AAanalysis" />
<meta property="og:description" content="Dive into the powerful capabilities of AAanalysis —a Python framework for sequence-based (i.e., alignment-free) and interpretable protein prediction. In this tutorial, we’ll focus on extracting phy..." />
<meta property="og:image:width" content="1146" />
<meta property="og:image:height" content="600" />
<meta property="og:image" content="/_images/social_previews/summary_generated_tutorial1_slow_start_8a675a32.png" />
<meta property="og:image:alt" content="Dive into the powerful capabilities of AAanalysis —a Python framework for sequence-based (i.e., alignment-free) and interpretable protein prediction. In this..." />
<meta name="description" content="Dive into the powerful capabilities of AAanalysis —a Python framework for sequence-based (i.e., alignment-free) and interpretable protein prediction. In this tutorial, we’ll focus on extracting phy..." />
<meta name="twitter:card" content="summary_large_image" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Slow Start with AAanalysis &mdash; AAanalysis</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="../_static/plot_directive.css" type="text/css" />
      <link rel="stylesheet" href="../_static/design-style.1e8bd061cd6da7fc9cf755528e8ffc24.min.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/style.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/notebook.css" type="text/css" />
    <link rel="shortcut icon" href="../_static/logo_small.png"/>
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
        <script src="../_static/clipboard.min.js"></script>
        <script src="../_static/copybutton.js"></script>
        <script src="../_static/design-tabs.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Plotting Prelude" href="plotting_prelude.html" />
    <link rel="prev" title="Quick Start with AAanalysis" href="tutorial1_quick_start.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #343131" >

          
          
          <a href="../index.html">
            
              <img src="../_static/logo_big_trans.png" class="logo" alt="Logo"/>
          </a>
              <div class="version">
                latest
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">OVERVIEW</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../index/introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../index/usage_principles.html">Usage Principles</a></li>
<li class="toctree-l1"><a class="reference internal" href="../index/CONTRIBUTING_COPY.html">Contributing</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">EXAMPLES</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../tutorials.html">Tutorials</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../tutorials.html#getting-started">Getting Started</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="tutorial1_quick_start.html">Quick Start with AAanalysis</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Slow Start with AAanalysis</a></li>
<li class="toctree-l3"><a class="reference internal" href="plotting_prelude.html">Plotting Prelude</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials.html#data-handling">Data Handling</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials.html#feature-engineering">Feature Engineering</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials.html#pu-learning">PU Learning</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials.html#explainable-ai">Explainable AI</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">REFERENCES</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../api.html">API</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../index/tables.html">Tables</a></li>
<li class="toctree-l1"><a class="reference internal" href="../index/references.html">References</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #343131" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">AAanalysis</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../tutorials.html">Tutorials</a></li>
      <li class="breadcrumb-item active">Slow Start with AAanalysis</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/breimanntools/aaanalysis/blob/master/generated/tutorial1_slow_start.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="slow-start-with-aaanalysis">
<h1>Slow Start with AAanalysis<a class="headerlink" href="#slow-start-with-aaanalysis" title="Permalink to this heading"></a></h1>
<p>Dive into the powerful capabilities of <strong>AAanalysis</strong>—a Python framework
for sequence-based (i.e., alignment-free) and interpretable protein
prediction. In this tutorial, we’ll focus on extracting physicochemical
features from protein sequences using the <em>Comparative Physical
Profiling (CPP)</em> algorithm and how these features can be harnessed for
binary classification tasks.</p>
<section id="what-you-will-learn">
<h2>What You Will Learn:<a class="headerlink" href="#what-you-will-learn" title="Permalink to this heading"></a></h2>
<ol class="arabic simple">
<li><p><strong>Loading Sequences and Scales</strong>: Load protein sequences and
selections of amino acid scales.</p></li>
<li><p><strong>Feature Engineering</strong>: Extract essential features using the
<code class="docutils literal notranslate"><span class="pre">AAclust</span></code> and <code class="docutils literal notranslate"><span class="pre">CPP</span></code> models.</p></li>
<li><p><strong>Protein Prediction</strong>: Make predictions using the RandomForest
model.</p></li>
<li><p><strong>Explainable AI</strong>: Interpret predictions at the group and individual
levels by combining <code class="docutils literal notranslate"><span class="pre">CPP</span></code> with <code class="docutils literal notranslate"><span class="pre">SHAP</span></code>.</p></li>
</ol>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="kn">import</span> <span class="nn">aaanalysis</span> <span class="k">as</span> <span class="nn">aa</span>
<span class="n">aa</span><span class="o">.</span><span class="n">options</span><span class="p">[</span><span class="s2">&quot;verbose&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">False</span>
</pre></div>
</div>
</section>
<section id="loading-sequences-and-scales">
<h2>1. Loading Sequences and Scales<a class="headerlink" href="#loading-sequences-and-scales" title="Permalink to this heading"></a></h2>
<p>With AAanalysis, you have access to numerous benchmark datasets for
protein sequence analysis. Using our γ-secretase substrates and
non-substrates dataset as a hands-on example, you can effortlessly
retrieve these datasets using the <code class="docutils literal notranslate"><span class="pre">aa.load_dataset()</span></code> function.
Furthermore, amino acid scales, predominantly from AAindex, along with
their hierarchical classification (known as <code class="docutils literal notranslate"><span class="pre">AAontology</span></code>), are
available at your fingertips with the <code class="docutils literal notranslate"><span class="pre">aa.load_scales()</span></code> function.</p>
<p>We now load the scales dataset and a dataset of 50 γ-secretase
substrates and non-substrates:</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_scales</span> <span class="o">=</span> <span class="n">aa</span><span class="o">.</span><span class="n">load_scales</span><span class="p">()</span>
<span class="n">df_seq</span> <span class="o">=</span> <span class="n">aa</span><span class="o">.</span><span class="n">load_dataset</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;DOM_GSEC&quot;</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="feature-engineering">
<h2>2. Feature Engineering<a class="headerlink" href="#feature-engineering" title="Permalink to this heading"></a></h2>
<p>The centerpiece of AAanalysis is the <code class="docutils literal notranslate"><span class="pre">CPP</span></code> model, which is supported
by <code class="docutils literal notranslate"><span class="pre">AAclust</span></code> for the pre-selection of amino acid scales.</p>
<section id="aaclust">
<h3>AAclust<a class="headerlink" href="#aaclust" title="Permalink to this heading"></a></h3>
<p>Since redundancy is an essential problem for machine learning tasks, the
<code class="docutils literal notranslate"><span class="pre">AAclust</span></code> object provides a lightweight wrapper for sklearn clustering
algorithms such as Agglomerative clustering. AAclust clusters a set of
scales and selects for each cluster the most representative scale (i.e.,
the scale closes to the cluster center). We will use AAclust to obtain a
set of 100 scales, as defined by the <code class="docutils literal notranslate"><span class="pre">n_clusters</span></code> parameters:</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">sklearn.cluster</span> <span class="kn">import</span> <span class="n">KMeans</span>

<span class="n">aac</span> <span class="o">=</span> <span class="n">aa</span><span class="o">.</span><span class="n">AAclust</span><span class="p">(</span><span class="n">model_class</span><span class="o">=</span><span class="n">KMeans</span><span class="p">)</span>
<span class="n">X</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">df_scales</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>
<span class="n">scales</span> <span class="o">=</span> <span class="n">aac</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">df_scales</span><span class="p">),</span> <span class="n">n_clusters</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span><span class="o">.</span><span class="n">medoid_names_</span>
<span class="n">df_scales</span> <span class="o">=</span> <span class="n">df_scales</span><span class="p">[</span><span class="n">scales</span><span class="p">]</span>
</pre></div>
</div>
</section>
<section id="comparative-physicochemical-profiling-cpp">
<h3>Comparative Physicochemical Profiling (CPP)<a class="headerlink" href="#comparative-physicochemical-profiling-cpp" title="Permalink to this heading"></a></h3>
<p>CPP is a sequence-based feature engineering algorithm. It aims at
identifying a set of features most discriminant between two sets of
sequences: the test set and the reference set. Supported by the
<code class="docutils literal notranslate"><span class="pre">SequenceFeature</span></code> object (<code class="docutils literal notranslate"><span class="pre">sf</span></code>), A CPP feature integrates:</p>
<ul class="simple">
<li><p><strong>Parts</strong>: Are combination of a target middle domain (TMD) and N- and
C-terminal adjacent regions (JMD-N and JMD-C, respectively), obtained
<code class="docutils literal notranslate"><span class="pre">sf.get_df_parts</span></code>.</p></li>
<li><p><strong>Splits</strong>: These Parts can be split into various continuous segments
or discontinuous patterns, specified <code class="docutils literal notranslate"><span class="pre">sf.get_split_kws()</span></code>.</p></li>
<li><p><strong>Scales</strong>: Sets of amino acid scales.</p></li>
</ul>
<p>We use SequenceFeature to obtain Parts and Splits. These together with
the Scales are used by CPP as input to identify the set of
characteristic features to discriminate between γ-secretase substrates
and non-substrates:</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sf</span> <span class="o">=</span> <span class="n">aa</span><span class="o">.</span><span class="n">SequenceFeature</span><span class="p">()</span>
<span class="n">df_parts</span> <span class="o">=</span> <span class="n">sf</span><span class="o">.</span><span class="n">get_df_parts</span><span class="p">(</span><span class="n">df_seq</span><span class="o">=</span><span class="n">df_seq</span><span class="p">,</span> <span class="n">list_parts</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;tmd&quot;</span><span class="p">])</span>
<span class="n">split_kws</span> <span class="o">=</span> <span class="n">sf</span><span class="o">.</span><span class="n">get_split_kws</span><span class="p">(</span><span class="n">n_split_max</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">split_types</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Segment&quot;</span><span class="p">])</span>
</pre></div>
</div>
<p>Running the CPP algorithm creates all Part, Split, Split combinations
and filters a selected maximum of non-redundant features. As a baseline
approach, we use CPP without filtering (<code class="docutils literal notranslate"><span class="pre">max_cor=1</span></code>) to compute the
average values for the 100 selected scales over the entire TMD sequences
(corresponding to the transmembrane domain of γ-secretase substrates and
non-substrates):</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Small set of CPP features (100 features are created)</span>
<span class="n">y</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">df_seq</span><span class="p">[</span><span class="s2">&quot;label&quot;</span><span class="p">])</span>
<span class="n">cpp</span> <span class="o">=</span> <span class="n">aa</span><span class="o">.</span><span class="n">CPP</span><span class="p">(</span><span class="n">df_scales</span><span class="o">=</span><span class="n">df_scales</span><span class="p">,</span> <span class="n">df_parts</span><span class="o">=</span><span class="n">df_parts</span><span class="p">,</span> <span class="n">split_kws</span><span class="o">=</span><span class="n">split_kws</span><span class="p">)</span>
<span class="n">df_feat</span> <span class="o">=</span> <span class="n">cpp</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">labels</span><span class="o">=</span><span class="n">y</span><span class="p">,</span> <span class="n">max_cor</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">aa</span><span class="o">.</span><span class="n">display_df</span><span class="p">(</span><span class="n">df</span><span class="o">=</span><span class="n">df_feat</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">))</span>
</pre></div>
</div>
<style type="text/css">
#T_c0b87 thead th {
  background-color: white;
  color: black;
}
#T_c0b87 tbody tr:nth-child(odd) {
  background-color: #f2f2f2;
}
#T_c0b87 tbody tr:nth-child(even) {
  background-color: white;
}
#T_c0b87 th {
  padding: 5px;
  white-space: nowrap;
}
#T_c0b87  td {
  padding: 5px;
  white-space: nowrap;
}
#T_c0b87 table {
  font-size: 12px;
}
</style>
<table id="T_c0b87" style='display:block; max-height: 300px; max-width: 100%; overflow-x: auto; overflow-y: auto;'>
  <thead>
    <tr>
      <th class="blank level0" >&nbsp;</th>
      <th id="T_c0b87_level0_col0" class="col_heading level0 col0" >feature</th>
      <th id="T_c0b87_level0_col1" class="col_heading level0 col1" >category</th>
      <th id="T_c0b87_level0_col2" class="col_heading level0 col2" >subcategory</th>
      <th id="T_c0b87_level0_col3" class="col_heading level0 col3" >scale_name</th>
      <th id="T_c0b87_level0_col4" class="col_heading level0 col4" >scale_description</th>
      <th id="T_c0b87_level0_col5" class="col_heading level0 col5" >abs_auc</th>
      <th id="T_c0b87_level0_col6" class="col_heading level0 col6" >abs_mean_dif</th>
      <th id="T_c0b87_level0_col7" class="col_heading level0 col7" >mean_dif</th>
      <th id="T_c0b87_level0_col8" class="col_heading level0 col8" >std_test</th>
      <th id="T_c0b87_level0_col9" class="col_heading level0 col9" >std_ref</th>
      <th id="T_c0b87_level0_col10" class="col_heading level0 col10" >p_val_mann_whitney</th>
      <th id="T_c0b87_level0_col11" class="col_heading level0 col11" >p_val_fdr_bh</th>
      <th id="T_c0b87_level0_col12" class="col_heading level0 col12" >positions</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th id="T_c0b87_level0_row0" class="row_heading level0 row0" >1</th>
      <td id="T_c0b87_row0_col0" class="data row0 col0" >TMD-Segment(1,1)-WOLR790101</td>
      <td id="T_c0b87_row0_col1" class="data row0 col1" >Polarity</td>
      <td id="T_c0b87_row0_col2" class="data row0 col2" >Hydrophobicity (surrounding)</td>
      <td id="T_c0b87_row0_col3" class="data row0 col3" >Hydration potential</td>
      <td id="T_c0b87_row0_col4" class="data row0 col4" >Hydrophobicity index (Wolfenden et al., 1979)</td>
      <td id="T_c0b87_row0_col5" class="data row0 col5" >0.246000</td>
      <td id="T_c0b87_row0_col6" class="data row0 col6" >0.032767</td>
      <td id="T_c0b87_row0_col7" class="data row0 col7" >0.032767</td>
      <td id="T_c0b87_row0_col8" class="data row0 col8" >0.028962</td>
      <td id="T_c0b87_row0_col9" class="data row0 col9" >0.037213</td>
      <td id="T_c0b87_row0_col10" class="data row0 col10" >0.000022</td>
      <td id="T_c0b87_row0_col11" class="data row0 col11" >0.002203</td>
      <td id="T_c0b87_row0_col12" class="data row0 col12" >11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30</td>
    </tr>
    <tr>
      <th id="T_c0b87_level0_row1" class="row_heading level0 row1" >2</th>
      <td id="T_c0b87_row1_col0" class="data row1 col0" >TMD-Segment(1,1)-ANDN920101</td>
      <td id="T_c0b87_row1_col1" class="data row1 col1" >Structure-Activity</td>
      <td id="T_c0b87_row1_col2" class="data row1 col2" >Backbone-dynamics (-CH)</td>
      <td id="T_c0b87_row1_col3" class="data row1 col3" >α-CH chemical shifts (backbone-dynamics)</td>
      <td id="T_c0b87_row1_col4" class="data row1 col4" >alpha-CH chemical shifts (Andersen et al., 1992)</td>
      <td id="T_c0b87_row1_col5" class="data row1 col5" >0.230000</td>
      <td id="T_c0b87_row1_col6" class="data row1 col6" >0.064845</td>
      <td id="T_c0b87_row1_col7" class="data row1 col7" >-0.064845</td>
      <td id="T_c0b87_row1_col8" class="data row1 col8" >0.072291</td>
      <td id="T_c0b87_row1_col9" class="data row1 col9" >0.077702</td>
      <td id="T_c0b87_row1_col10" class="data row1 col10" >0.000076</td>
      <td id="T_c0b87_row1_col11" class="data row1 col11" >0.002528</td>
      <td id="T_c0b87_row1_col12" class="data row1 col12" >11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30</td>
    </tr>
    <tr>
      <th id="T_c0b87_level0_row2" class="row_heading level0 row2" >3</th>
      <td id="T_c0b87_row2_col0" class="data row2 col0" >TMD-Segment(1,1)-KOEH090103</td>
      <td id="T_c0b87_row2_col1" class="data row2 col1" >Polarity</td>
      <td id="T_c0b87_row2_col2" class="data row2 col2" >Hydrophilicity</td>
      <td id="T_c0b87_row2_col3" class="data row2 col3" >Polarity (hydrophilicity)</td>
      <td id="T_c0b87_row2_col4" class="data row2 col4" >Hydrophobicity scale (Eisenberg-Weiss, 1984), Inverted to match the direction of transfer from water to bilyer by Koehler et al. (2009)</td>
      <td id="T_c0b87_row2_col5" class="data row2 col5" >0.227000</td>
      <td id="T_c0b87_row2_col6" class="data row2 col6" >0.023491</td>
      <td id="T_c0b87_row2_col7" class="data row2 col7" >-0.023491</td>
      <td id="T_c0b87_row2_col8" class="data row2 col8" >0.024560</td>
      <td id="T_c0b87_row2_col9" class="data row2 col9" >0.032585</td>
      <td id="T_c0b87_row2_col10" class="data row2 col10" >0.000094</td>
      <td id="T_c0b87_row2_col11" class="data row2 col11" >0.002528</td>
      <td id="T_c0b87_row2_col12" class="data row2 col12" >11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30</td>
    </tr>
    <tr>
      <th id="T_c0b87_level0_row3" class="row_heading level0 row3" >4</th>
      <td id="T_c0b87_row3_col0" class="data row3 col0" >TMD-Segment(1,1)-LINS030104</td>
      <td id="T_c0b87_row3_col1" class="data row3 col1" >ASA/Volume</td>
      <td id="T_c0b87_row3_col2" class="data row3 col2" >Accessible surface area (ASA)</td>
      <td id="T_c0b87_row3_col3" class="data row3 col3" >ASA (folded protein)</td>
      <td id="T_c0b87_row3_col4" class="data row3 col4" >Total median accessible surfaces of whole residues from folded proteins, expressed in Å²  (Lins et al., 2003)</td>
      <td id="T_c0b87_row3_col5" class="data row3 col5" >0.223000</td>
      <td id="T_c0b87_row3_col6" class="data row3 col6" >0.025527</td>
      <td id="T_c0b87_row3_col7" class="data row3 col7" >-0.025527</td>
      <td id="T_c0b87_row3_col8" class="data row3 col8" >0.024048</td>
      <td id="T_c0b87_row3_col9" class="data row3 col9" >0.033495</td>
      <td id="T_c0b87_row3_col10" class="data row3 col10" >0.000120</td>
      <td id="T_c0b87_row3_col11" class="data row3 col11" >0.002528</td>
      <td id="T_c0b87_row3_col12" class="data row3 col12" >11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30</td>
    </tr>
    <tr>
      <th id="T_c0b87_level0_row4" class="row_heading level0 row4" >5</th>
      <td id="T_c0b87_row4_col0" class="data row4 col0" >TMD-Segment(1,1)-LINS030106</td>
      <td id="T_c0b87_row4_col1" class="data row4 col1" >ASA/Volume</td>
      <td id="T_c0b87_row4_col2" class="data row4 col2" >Accessible surface area (ASA)</td>
      <td id="T_c0b87_row4_col3" class="data row4 col3" >Hydrophilic ASA (folded proteins)</td>
      <td id="T_c0b87_row4_col4" class="data row4 col4" >Hydrophilic median accessible surfaces of whole residues from folded proteins, expressed in Å²  (Lins et al., 2003)</td>
      <td id="T_c0b87_row4_col5" class="data row4 col5" >0.222000</td>
      <td id="T_c0b87_row4_col6" class="data row4 col6" >0.024464</td>
      <td id="T_c0b87_row4_col7" class="data row4 col7" >-0.024464</td>
      <td id="T_c0b87_row4_col8" class="data row4 col8" >0.023645</td>
      <td id="T_c0b87_row4_col9" class="data row4 col9" >0.032067</td>
      <td id="T_c0b87_row4_col10" class="data row4 col10" >0.000136</td>
      <td id="T_c0b87_row4_col11" class="data row4 col11" >0.002528</td>
      <td id="T_c0b87_row4_col12" class="data row4 col12" >11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30</td>
    </tr>
    <tr>
      <th id="T_c0b87_level0_row5" class="row_heading level0 row5" >6</th>
      <td id="T_c0b87_row5_col0" class="data row5 col0" >TMD-Segment(1,1)-CHOC760104</td>
      <td id="T_c0b87_row5_col1" class="data row5 col1" >ASA/Volume</td>
      <td id="T_c0b87_row5_col2" class="data row5 col2" >Buried</td>
      <td id="T_c0b87_row5_col3" class="data row5 col3" >Buried</td>
      <td id="T_c0b87_row5_col4" class="data row5 col4" >Proportion of residues 100% buried (Chothia, 1976)</td>
      <td id="T_c0b87_row5_col5" class="data row5 col5" >0.220000</td>
      <td id="T_c0b87_row5_col6" class="data row5 col6" >0.046652</td>
      <td id="T_c0b87_row5_col7" class="data row5 col7" >0.046652</td>
      <td id="T_c0b87_row5_col8" class="data row5 col8" >0.046731</td>
      <td id="T_c0b87_row5_col9" class="data row5 col9" >0.062379</td>
      <td id="T_c0b87_row5_col10" class="data row5 col10" >0.000152</td>
      <td id="T_c0b87_row5_col11" class="data row5 col11" >0.002528</td>
      <td id="T_c0b87_row5_col12" class="data row5 col12" >11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30</td>
    </tr>
    <tr>
      <th id="T_c0b87_level0_row6" class="row_heading level0 row6" >7</th>
      <td id="T_c0b87_row6_col0" class="data row6 col0" >TMD-Segment(1,1)-JURD980101</td>
      <td id="T_c0b87_row6_col1" class="data row6 col1" >Polarity</td>
      <td id="T_c0b87_row6_col2" class="data row6 col2" >Hydrophobicity</td>
      <td id="T_c0b87_row6_col3" class="data row6 col3" >Hydrophobicity</td>
      <td id="T_c0b87_row6_col4" class="data row6 col4" >Modified Kyte-Doolittle hydrophobicity scale (Juretic et al., 1998)</td>
      <td id="T_c0b87_row6_col5" class="data row6 col5" >0.209000</td>
      <td id="T_c0b87_row6_col6" class="data row6 col6" >0.038451</td>
      <td id="T_c0b87_row6_col7" class="data row6 col7" >0.038451</td>
      <td id="T_c0b87_row6_col8" class="data row6 col8" >0.044812</td>
      <td id="T_c0b87_row6_col9" class="data row6 col9" >0.055053</td>
      <td id="T_c0b87_row6_col10" class="data row6 col10" >0.000324</td>
      <td id="T_c0b87_row6_col11" class="data row6 col11" >0.004388</td>
      <td id="T_c0b87_row6_col12" class="data row6 col12" >11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30</td>
    </tr>
    <tr>
      <th id="T_c0b87_level0_row7" class="row_heading level0 row7" >8</th>
      <td id="T_c0b87_row7_col0" class="data row7 col0" >TMD-Segment(1,1)-FUKS010106</td>
      <td id="T_c0b87_row7_col1" class="data row7 col1" >Composition</td>
      <td id="T_c0b87_row7_col2" class="data row7 col2" >Membrane proteins (MPs)</td>
      <td id="T_c0b87_row7_col3" class="data row7 col3" >Proteins of mesophiles (INT)</td>
      <td id="T_c0b87_row7_col4" class="data row7 col4" >Interior composition of amino acids in intracellular proteins of mesophiles [%] (Fukuchi-Nishikawa, 2001)</td>
      <td id="T_c0b87_row7_col5" class="data row7 col5" >0.206000</td>
      <td id="T_c0b87_row7_col6" class="data row7 col6" >0.058909</td>
      <td id="T_c0b87_row7_col7" class="data row7 col7" >0.058909</td>
      <td id="T_c0b87_row7_col8" class="data row7 col8" >0.068070</td>
      <td id="T_c0b87_row7_col9" class="data row7 col9" >0.081967</td>
      <td id="T_c0b87_row7_col10" class="data row7 col10" >0.000380</td>
      <td id="T_c0b87_row7_col11" class="data row7 col11" >0.004388</td>
      <td id="T_c0b87_row7_col12" class="data row7 col12" >11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30</td>
    </tr>
    <tr>
      <th id="T_c0b87_level0_row8" class="row_heading level0 row8" >9</th>
      <td id="T_c0b87_row8_col0" class="data row8 col0" >TMD-Segment(1,1)-KOEH090102</td>
      <td id="T_c0b87_row8_col1" class="data row8 col1" >Polarity</td>
      <td id="T_c0b87_row8_col2" class="data row8 col2" >Hydrophilicity</td>
      <td id="T_c0b87_row8_col3" class="data row8 col3" >Polarity (hydrophilicity)</td>
      <td id="T_c0b87_row8_col4" class="data row8 col4" >Hydrophobicity scale (Hessa et al., 2007), Cited by Koehler et al. (2009)</td>
      <td id="T_c0b87_row8_col5" class="data row8 col5" >0.206000</td>
      <td id="T_c0b87_row8_col6" class="data row8 col6" >0.025848</td>
      <td id="T_c0b87_row8_col7" class="data row8 col7" >-0.025848</td>
      <td id="T_c0b87_row8_col8" class="data row8 col8" >0.028664</td>
      <td id="T_c0b87_row8_col9" class="data row8 col9" >0.039067</td>
      <td id="T_c0b87_row8_col10" class="data row8 col10" >0.000395</td>
      <td id="T_c0b87_row8_col11" class="data row8 col11" >0.004388</td>
      <td id="T_c0b87_row8_col12" class="data row8 col12" >11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30</td>
    </tr>
    <tr>
      <th id="T_c0b87_level0_row9" class="row_heading level0 row9" >10</th>
      <td id="T_c0b87_row9_col0" class="data row9 col0" >TMD-Segment(1,1)-VELV850101</td>
      <td id="T_c0b87_row9_col1" class="data row9 col1" >Energy</td>
      <td id="T_c0b87_row9_col2" class="data row9 col2" >Electron-ion interaction pot.</td>
      <td id="T_c0b87_row9_col3" class="data row9 col3" >Electron-ion interaction potential</td>
      <td id="T_c0b87_row9_col4" class="data row9 col4" >Electron-ion interaction potential (Veljkovic et al., 1985)</td>
      <td id="T_c0b87_row9_col5" class="data row9 col5" >0.203000</td>
      <td id="T_c0b87_row9_col6" class="data row9 col6" >0.045725</td>
      <td id="T_c0b87_row9_col7" class="data row9 col7" >-0.045725</td>
      <td id="T_c0b87_row9_col8" class="data row9 col8" >0.059791</td>
      <td id="T_c0b87_row9_col9" class="data row9 col9" >0.068804</td>
      <td id="T_c0b87_row9_col10" class="data row9 col10" >0.000480</td>
      <td id="T_c0b87_row9_col11" class="data row9 col11" >0.004687</td>
      <td id="T_c0b87_row9_col12" class="data row9 col12" >11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30</td>
    </tr>
  </tbody>
</table></section>
</section>
<section id="protein-prediction">
<h2>3. Protein Prediction<a class="headerlink" href="#protein-prediction" title="Permalink to this heading"></a></h2>
<p>A feature matrix from a given set of CPP features can be created using
<code class="docutils literal notranslate"><span class="pre">sf.feat_matrix</span></code> and used for machine learning:</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">sklearn.ensemble</span> <span class="kn">import</span> <span class="n">RandomForestClassifier</span>
<span class="kn">from</span> <span class="nn">sklearn.model_selection</span> <span class="kn">import</span> <span class="n">cross_val_score</span>

<span class="n">X</span> <span class="o">=</span> <span class="n">sf</span><span class="o">.</span><span class="n">feature_matrix</span><span class="p">(</span><span class="n">df_parts</span><span class="o">=</span><span class="n">df_parts</span><span class="p">,</span> <span class="n">features</span><span class="o">=</span><span class="n">df_feat</span><span class="p">[</span><span class="s2">&quot;feature&quot;</span><span class="p">])</span>
<span class="n">rf</span> <span class="o">=</span> <span class="n">RandomForestClassifier</span><span class="p">()</span>
<span class="n">cv_base</span> <span class="o">=</span> <span class="n">cross_val_score</span><span class="p">(</span><span class="n">rf</span><span class="p">,</span> <span class="n">X</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">scoring</span><span class="o">=</span><span class="s2">&quot;accuracy&quot;</span><span class="p">,</span> <span class="n">cv</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Mean accuracy of </span><span class="si">{</span><span class="nb">round</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">cv_base</span><span class="p">),</span><span class="w"> </span><span class="mi">2</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Mean</span> <span class="n">accuracy</span> <span class="n">of</span> <span class="mf">0.56</span>
</pre></div>
</div>
<p>CPP uses 330 Splits and 3 Parts by default, yielding around 100.000
features for 100 Scales. Creating and filtering all these
Part-Split-Scale combinations will take a little time (~1.5 minutes for
this example on an i7-10510U (4 cores, 8 threads) with multiprocessing),
but improve prediction performance:</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># CPP features with default splits and parts (around 100.000 features)</span>
<span class="n">df_parts</span> <span class="o">=</span> <span class="n">sf</span><span class="o">.</span><span class="n">get_df_parts</span><span class="p">(</span><span class="n">df_seq</span><span class="o">=</span><span class="n">df_seq</span><span class="p">)</span>
<span class="n">cpp</span> <span class="o">=</span> <span class="n">aa</span><span class="o">.</span><span class="n">CPP</span><span class="p">(</span><span class="n">df_scales</span><span class="o">=</span><span class="n">df_scales</span><span class="p">,</span> <span class="n">df_parts</span><span class="o">=</span><span class="n">df_parts</span><span class="p">)</span>
<span class="n">df_feat</span> <span class="o">=</span> <span class="n">cpp</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">labels</span><span class="o">=</span><span class="n">y</span><span class="p">)</span>
<span class="n">aa</span><span class="o">.</span><span class="n">display_df</span><span class="p">(</span><span class="n">df</span><span class="o">=</span><span class="n">df_feat</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">))</span>

<span class="n">X</span> <span class="o">=</span> <span class="n">sf</span><span class="o">.</span><span class="n">feature_matrix</span><span class="p">(</span><span class="n">df_parts</span><span class="o">=</span><span class="n">df_parts</span><span class="p">,</span> <span class="n">features</span><span class="o">=</span><span class="n">df_feat</span><span class="p">[</span><span class="s2">&quot;feature&quot;</span><span class="p">])</span>
<span class="n">rf</span> <span class="o">=</span> <span class="n">RandomForestClassifier</span><span class="p">()</span>
<span class="n">cv</span> <span class="o">=</span> <span class="n">cross_val_score</span><span class="p">(</span><span class="n">rf</span><span class="p">,</span> <span class="n">X</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">scoring</span><span class="o">=</span><span class="s2">&quot;accuracy&quot;</span><span class="p">,</span> <span class="n">cv</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Mean accuracy of </span><span class="si">{</span><span class="nb">round</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">cv</span><span class="p">),</span><span class="w"> </span><span class="mi">2</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># Plot comparison of two approaches</span>
<span class="n">aa</span><span class="o">.</span><span class="n">plot_settings</span><span class="p">()</span>
<span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;Baseline&quot;</span><span class="p">:</span> <span class="n">cv_base</span><span class="p">,</span> <span class="s2">&quot;CPP&quot;</span><span class="p">:</span> <span class="n">cv</span><span class="p">}),</span> <span class="n">palette</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;tab:blue&quot;</span><span class="p">,</span> <span class="s2">&quot;tab:red&quot;</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Mean accuracy&quot;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">aa</span><span class="o">.</span><span class="n">plot_gcfs</span><span class="p">()</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Comparison of Feature Engineering Methods&quot;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">aa</span><span class="o">.</span><span class="n">plot_gcfs</span><span class="p">()</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">despine</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
<style type="text/css">
#T_ce980 thead th {
  background-color: white;
  color: black;
}
#T_ce980 tbody tr:nth-child(odd) {
  background-color: #f2f2f2;
}
#T_ce980 tbody tr:nth-child(even) {
  background-color: white;
}
#T_ce980 th {
  padding: 5px;
  white-space: nowrap;
}
#T_ce980  td {
  padding: 5px;
  white-space: nowrap;
}
#T_ce980 table {
  font-size: 12px;
}
</style>
<table id="T_ce980" style='display:block; max-height: 300px; max-width: 100%; overflow-x: auto; overflow-y: auto;'>
  <thead>
    <tr>
      <th class="blank level0" >&nbsp;</th>
      <th id="T_ce980_level0_col0" class="col_heading level0 col0" >feature</th>
      <th id="T_ce980_level0_col1" class="col_heading level0 col1" >category</th>
      <th id="T_ce980_level0_col2" class="col_heading level0 col2" >subcategory</th>
      <th id="T_ce980_level0_col3" class="col_heading level0 col3" >scale_name</th>
      <th id="T_ce980_level0_col4" class="col_heading level0 col4" >scale_description</th>
      <th id="T_ce980_level0_col5" class="col_heading level0 col5" >abs_auc</th>
      <th id="T_ce980_level0_col6" class="col_heading level0 col6" >abs_mean_dif</th>
      <th id="T_ce980_level0_col7" class="col_heading level0 col7" >mean_dif</th>
      <th id="T_ce980_level0_col8" class="col_heading level0 col8" >std_test</th>
      <th id="T_ce980_level0_col9" class="col_heading level0 col9" >std_ref</th>
      <th id="T_ce980_level0_col10" class="col_heading level0 col10" >p_val_mann_whitney</th>
      <th id="T_ce980_level0_col11" class="col_heading level0 col11" >p_val_fdr_bh</th>
      <th id="T_ce980_level0_col12" class="col_heading level0 col12" >positions</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th id="T_ce980_level0_row0" class="row_heading level0 row0" >1</th>
      <td id="T_ce980_row0_col0" class="data row0 col0" >TMD_C_JMD_C-Segment(2,3)-QIAN880106</td>
      <td id="T_ce980_row0_col1" class="data row0 col1" >Conformation</td>
      <td id="T_ce980_row0_col2" class="data row0 col2" >α-helix</td>
      <td id="T_ce980_row0_col3" class="data row0 col3" >α-helix (middle)</td>
      <td id="T_ce980_row0_col4" class="data row0 col4" >Weights for alpha-helix at the window position of -1 (Qian-Sejnowski, 1988)</td>
      <td id="T_ce980_row0_col5" class="data row0 col5" >0.387000</td>
      <td id="T_ce980_row0_col6" class="data row0 col6" >0.121446</td>
      <td id="T_ce980_row0_col7" class="data row0 col7" >0.121446</td>
      <td id="T_ce980_row0_col8" class="data row0 col8" >0.069196</td>
      <td id="T_ce980_row0_col9" class="data row0 col9" >0.085013</td>
      <td id="T_ce980_row0_col10" class="data row0 col10" >0.000000</td>
      <td id="T_ce980_row0_col11" class="data row0 col11" >0.000000</td>
      <td id="T_ce980_row0_col12" class="data row0 col12" >27,28,29,30,31,32,33</td>
    </tr>
    <tr>
      <th id="T_ce980_level0_row1" class="row_heading level0 row1" >2</th>
      <td id="T_ce980_row1_col0" class="data row1 col0" >TMD_C_JMD_C-Segment(5,7)-ONEK900101</td>
      <td id="T_ce980_row1_col1" class="data row1 col1" >Others</td>
      <td id="T_ce980_row1_col2" class="data row1 col2" >Unclassified (Others)</td>
      <td id="T_ce980_row1_col3" class="data row1 col3" >ΔG values in peptides</td>
      <td id="T_ce980_row1_col4" class="data row1 col4" >Delta G values for the peptides extrapolated to 0 M urea (O'Neil-DeGrado, 1990)</td>
      <td id="T_ce980_row1_col5" class="data row1 col5" >0.373000</td>
      <td id="T_ce980_row1_col6" class="data row1 col6" >0.115273</td>
      <td id="T_ce980_row1_col7" class="data row1 col7" >0.115273</td>
      <td id="T_ce980_row1_col8" class="data row1 col8" >0.065969</td>
      <td id="T_ce980_row1_col9" class="data row1 col9" >0.112756</td>
      <td id="T_ce980_row1_col10" class="data row1 col10" >0.000000</td>
      <td id="T_ce980_row1_col11" class="data row1 col11" >0.000000</td>
      <td id="T_ce980_row1_col12" class="data row1 col12" >32,33,34</td>
    </tr>
    <tr>
      <th id="T_ce980_level0_row2" class="row_heading level0 row2" >3</th>
      <td id="T_ce980_row2_col0" class="data row2 col0" >TMD_C_JMD_C-Segment(5,7)-LEVM760105</td>
      <td id="T_ce980_row2_col1" class="data row2 col1" >Shape</td>
      <td id="T_ce980_row2_col2" class="data row2 col2" >Side chain length</td>
      <td id="T_ce980_row2_col3" class="data row2 col3" >Side chain length</td>
      <td id="T_ce980_row2_col4" class="data row2 col4" >Radius of gyration of side chain (Levitt, 1976)</td>
      <td id="T_ce980_row2_col5" class="data row2 col5" >0.360000</td>
      <td id="T_ce980_row2_col6" class="data row2 col6" >0.250473</td>
      <td id="T_ce980_row2_col7" class="data row2 col7" >0.250473</td>
      <td id="T_ce980_row2_col8" class="data row2 col8" >0.148804</td>
      <td id="T_ce980_row2_col9" class="data row2 col9" >0.174650</td>
      <td id="T_ce980_row2_col10" class="data row2 col10" >0.000000</td>
      <td id="T_ce980_row2_col11" class="data row2 col11" >0.000000</td>
      <td id="T_ce980_row2_col12" class="data row2 col12" >32,33,34</td>
    </tr>
    <tr>
      <th id="T_ce980_level0_row3" class="row_heading level0 row3" >4</th>
      <td id="T_ce980_row3_col0" class="data row3 col0" >TMD_C_JMD_C-Pattern(N,5,8,12,15)-QIAN880106</td>
      <td id="T_ce980_row3_col1" class="data row3 col1" >Conformation</td>
      <td id="T_ce980_row3_col2" class="data row3 col2" >α-helix</td>
      <td id="T_ce980_row3_col3" class="data row3 col3" >α-helix (middle)</td>
      <td id="T_ce980_row3_col4" class="data row3 col4" >Weights for alpha-helix at the window position of -1 (Qian-Sejnowski, 1988)</td>
      <td id="T_ce980_row3_col5" class="data row3 col5" >0.358000</td>
      <td id="T_ce980_row3_col6" class="data row3 col6" >0.144860</td>
      <td id="T_ce980_row3_col7" class="data row3 col7" >0.144860</td>
      <td id="T_ce980_row3_col8" class="data row3 col8" >0.079321</td>
      <td id="T_ce980_row3_col9" class="data row3 col9" >0.117515</td>
      <td id="T_ce980_row3_col10" class="data row3 col10" >0.000000</td>
      <td id="T_ce980_row3_col11" class="data row3 col11" >0.000000</td>
      <td id="T_ce980_row3_col12" class="data row3 col12" >25,28,32,35</td>
    </tr>
    <tr>
      <th id="T_ce980_level0_row4" class="row_heading level0 row4" >5</th>
      <td id="T_ce980_row4_col0" class="data row4 col0" >TMD_C_JMD_C-Segment(4,5)-KLEP840101</td>
      <td id="T_ce980_row4_col1" class="data row4 col1" >Energy</td>
      <td id="T_ce980_row4_col2" class="data row4 col2" >Charge</td>
      <td id="T_ce980_row4_col3" class="data row4 col3" >Charge</td>
      <td id="T_ce980_row4_col4" class="data row4 col4" >Net charge (Klein et al., 1984)</td>
      <td id="T_ce980_row4_col5" class="data row4 col5" >0.354000</td>
      <td id="T_ce980_row4_col6" class="data row4 col6" >0.192500</td>
      <td id="T_ce980_row4_col7" class="data row4 col7" >0.192500</td>
      <td id="T_ce980_row4_col8" class="data row4 col8" >0.111915</td>
      <td id="T_ce980_row4_col9" class="data row4 col9" >0.127009</td>
      <td id="T_ce980_row4_col10" class="data row4 col10" >0.000000</td>
      <td id="T_ce980_row4_col11" class="data row4 col11" >0.000000</td>
      <td id="T_ce980_row4_col12" class="data row4 col12" >33,34,35,36</td>
    </tr>
    <tr>
      <th id="T_ce980_level0_row5" class="row_heading level0 row5" >6</th>
      <td id="T_ce980_row5_col0" class="data row5 col0" >TMD_C_JMD_C-Pattern(N,4,8,12,15)-MUNV940102</td>
      <td id="T_ce980_row5_col1" class="data row5 col1" >Energy</td>
      <td id="T_ce980_row5_col2" class="data row5 col2" >Free energy (folding)</td>
      <td id="T_ce980_row5_col3" class="data row5 col3" >Free energy (α-helix)</td>
      <td id="T_ce980_row5_col4" class="data row5 col4" >Free energy in alpha-helical region (Munoz-Serrano, 1994)</td>
      <td id="T_ce980_row5_col5" class="data row5 col5" >0.353000</td>
      <td id="T_ce980_row5_col6" class="data row5 col6" >0.119820</td>
      <td id="T_ce980_row5_col7" class="data row5 col7" >-0.119820</td>
      <td id="T_ce980_row5_col8" class="data row5 col8" >0.065320</td>
      <td id="T_ce980_row5_col9" class="data row5 col9" >0.098536</td>
      <td id="T_ce980_row5_col10" class="data row5 col10" >0.000000</td>
      <td id="T_ce980_row5_col11" class="data row5 col11" >0.000001</td>
      <td id="T_ce980_row5_col12" class="data row5 col12" >24,28,32,35</td>
    </tr>
    <tr>
      <th id="T_ce980_level0_row6" class="row_heading level0 row6" >7</th>
      <td id="T_ce980_row6_col0" class="data row6 col0" >TMD_C_JMD_C-Segment(5,7)-HUTJ700102</td>
      <td id="T_ce980_row6_col1" class="data row6 col1" >Energy</td>
      <td id="T_ce980_row6_col2" class="data row6 col2" >Entropy</td>
      <td id="T_ce980_row6_col3" class="data row6 col3" >Entropy</td>
      <td id="T_ce980_row6_col4" class="data row6 col4" >Absolute entropy (Hutchens, 1970)</td>
      <td id="T_ce980_row6_col5" class="data row6 col5" >0.343000</td>
      <td id="T_ce980_row6_col6" class="data row6 col6" >0.229253</td>
      <td id="T_ce980_row6_col7" class="data row6 col7" >0.229253</td>
      <td id="T_ce980_row6_col8" class="data row6 col8" >0.143391</td>
      <td id="T_ce980_row6_col9" class="data row6 col9" >0.171184</td>
      <td id="T_ce980_row6_col10" class="data row6 col10" >0.000000</td>
      <td id="T_ce980_row6_col11" class="data row6 col11" >0.000001</td>
      <td id="T_ce980_row6_col12" class="data row6 col12" >32,33,34</td>
    </tr>
    <tr>
      <th id="T_ce980_level0_row7" class="row_heading level0 row7" >8</th>
      <td id="T_ce980_row7_col0" class="data row7 col0" >TMD-Segment(12,13)-ROBB760113</td>
      <td id="T_ce980_row7_col1" class="data row7 col1" >Conformation</td>
      <td id="T_ce980_row7_col2" class="data row7 col2" >β-turn</td>
      <td id="T_ce980_row7_col3" class="data row7 col3" >β-turn</td>
      <td id="T_ce980_row7_col4" class="data row7 col4" >Information measure for loop (Robson-Suzuki, 1976)</td>
      <td id="T_ce980_row7_col5" class="data row7 col5" >0.337000</td>
      <td id="T_ce980_row7_col6" class="data row7 col6" >0.319440</td>
      <td id="T_ce980_row7_col7" class="data row7 col7" >-0.319440</td>
      <td id="T_ce980_row7_col8" class="data row7 col8" >0.175203</td>
      <td id="T_ce980_row7_col9" class="data row7 col9" >0.255754</td>
      <td id="T_ce980_row7_col10" class="data row7 col10" >0.000000</td>
      <td id="T_ce980_row7_col11" class="data row7 col11" >0.000001</td>
      <td id="T_ce980_row7_col12" class="data row7 col12" >27,28</td>
    </tr>
    <tr>
      <th id="T_ce980_level0_row8" class="row_heading level0 row8" >9</th>
      <td id="T_ce980_row8_col0" class="data row8 col0" >TMD_C_JMD_C-Segment(4,5)-RICJ880113</td>
      <td id="T_ce980_row8_col1" class="data row8 col1" >Conformation</td>
      <td id="T_ce980_row8_col2" class="data row8 col2" >α-helix (C-cap)</td>
      <td id="T_ce980_row8_col3" class="data row8 col3" >α-helix (C-terminal, inside)</td>
      <td id="T_ce980_row8_col4" class="data row8 col4" >Relative preference value at C2 (Richardson-Richardson, 1988)</td>
      <td id="T_ce980_row8_col5" class="data row8 col5" >0.336000</td>
      <td id="T_ce980_row8_col6" class="data row8 col6" >0.223765</td>
      <td id="T_ce980_row8_col7" class="data row8 col7" >0.223765</td>
      <td id="T_ce980_row8_col8" class="data row8 col8" >0.133513</td>
      <td id="T_ce980_row8_col9" class="data row8 col9" >0.178217</td>
      <td id="T_ce980_row8_col10" class="data row8 col10" >0.000000</td>
      <td id="T_ce980_row8_col11" class="data row8 col11" >0.000001</td>
      <td id="T_ce980_row8_col12" class="data row8 col12" >33,34,35,36</td>
    </tr>
    <tr>
      <th id="T_ce980_level0_row9" class="row_heading level0 row9" >10</th>
      <td id="T_ce980_row9_col0" class="data row9 col0" >TMD_C_JMD_C-Pattern(C,4,8)-JURD980101</td>
      <td id="T_ce980_row9_col1" class="data row9 col1" >Polarity</td>
      <td id="T_ce980_row9_col2" class="data row9 col2" >Hydrophobicity</td>
      <td id="T_ce980_row9_col3" class="data row9 col3" >Hydrophobicity</td>
      <td id="T_ce980_row9_col4" class="data row9 col4" >Modified Kyte-Doolittle hydrophobicity scale (Juretic et al., 1998)</td>
      <td id="T_ce980_row9_col5" class="data row9 col5" >0.329000</td>
      <td id="T_ce980_row9_col6" class="data row9 col6" >0.264720</td>
      <td id="T_ce980_row9_col7" class="data row9 col7" >-0.264720</td>
      <td id="T_ce980_row9_col8" class="data row9 col8" >0.141666</td>
      <td id="T_ce980_row9_col9" class="data row9 col9" >0.233134</td>
      <td id="T_ce980_row9_col10" class="data row9 col10" >0.000000</td>
      <td id="T_ce980_row9_col11" class="data row9 col11" >0.000001</td>
      <td id="T_ce980_row9_col12" class="data row9 col12" >33,37</td>
    </tr>
  </tbody>
</table><div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Mean</span> <span class="n">accuracy</span> <span class="n">of</span> <span class="mf">0.9</span>
</pre></div>
</div>
<img alt="../_images/tutorial1_slow_start_1_output_13_2.png" src="../_images/tutorial1_slow_start_1_output_13_2.png" />
</section>
<section id="explainable-ai">
<h2>4. Explainable AI<a class="headerlink" href="#explainable-ai" title="Permalink to this heading"></a></h2>
<section id="explainable-ai-on-group-level">
<h3>Explainable AI on group level<a class="headerlink" href="#explainable-ai-on-group-level" title="Permalink to this heading"></a></h3>
</section>
<section id="explainable-ai-on-individual-level">
<h3>Explainable AI on individual level<a class="headerlink" href="#explainable-ai-on-individual-level" title="Permalink to this heading"></a></h3>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="tutorial1_quick_start.html" class="btn btn-neutral float-left" title="Quick Start with AAanalysis" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="plotting_prelude.html" class="btn btn-neutral float-right" title="Plotting Prelude" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Stephan Breimann.</p>
  </div>

   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>